private/bufpkg/bufconnect/interceptors.go

```diff
@@ -43,64 +44,120 @@ func NewSetCLIVersionInterceptor(version string) connect.UnaryInterceptorFunc {
-// a client.
-func NewAuthorizationInterceptorProviderWithToken(token string) func(string) connect.UnaryInterceptorFunc {
-	return func(address string) connect.UnaryInterceptorFunc {
-		interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
-			return connect.UnaryFunc(func(
-				ctx context.Context,
-				req connect.AnyRequest,
-			) (connect.AnyResponse, error) {
-				if token != "" {
-					req.Header().Set(AuthenticationHeader, AuthenticationTokenPrefix+token)
-				}
-				return next(ctx, req)
-			})
+// SetAuthTokenOption is an option for NewAuthorizationInterceptorProvider
+type SetAuthTokenOption func(connect.AnyRequest, context.Context, connect.UnaryFunc, string) (connect.AnyResponse, error)
+
+// SetAuthTokenWithProvidedToken returns a new SetAuthTokenOption that will set the
+// provided token into the request header This is used for registry providers where the token is known at provider
+// creation (i.e. when logging in and explicitly pasting a token into stdin
+func SetAuthTokenWithProvidedToken(token string) SetAuthTokenOption {
+	return func(req connect.AnyRequest, ctx context.Context, next connect.UnaryFunc, address string) (connect.AnyResponse, error) {
+		if token != "" {
+			req.Header().Set(AuthenticationHeader, AuthenticationTokenPrefix+token)
 		}
-		return interceptor
+		return next(ctx, req)
+	}
+}
+
+// SetAuthTokenWithAddress returns a new SetAuthTokenOption that will loop up an auth token
+// and set it into the request header
+func SetAuthTokenWithAddress(container app.EnvContainer) SetAuthTokenOption {
+	return func(req connect.AnyRequest, ctx context.Context, next connect.UnaryFunc, address string) (connect.AnyResponse, error) {
+		envKey := tokenEnvKey
+		token := container.Env(envKey)
+		authorizationToken := ""
+		if token == "" {
+			envKey = ""
+			machine, err := netrc.GetMachineForName(container, address)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read server password from netrc: %w", err)
+			}
+			if machine != nil {
+				authorizationToken = machine.Password()
+			}
+		}
+		if token != "" {
+			tokenSet, err := newTokenSetFromString(token)
+			if err != nil {
+				return nil, err
+			}
+			_, authorizationToken = tokenSet.getRemoteUsernameAndToken(address)
+		}
+		if authorizationToken != "" {
+			req.Header().Set(AuthenticationHeader, AuthenticationTokenPrefix+authorizationToken)
+		}
+		response, err := next(ctx, req)
+		if err != nil {
+			err = &ErrAuth{cause: err, tokenEnvKey: envKey}
+		}
+		return response, err
+	}
+}
+
+type tokenSet struct {
+	bufToken        string
+	remoteUsernames map[string]string
+	remoteTokens    map[string]string
+}
+
+func newTokenSetFromString(token string) (*tokenSet, error) {
+	tokenSet := &tokenSet{
+		remoteUsernames: make(map[string]string),
+		remoteTokens:    make(map[string]string),
+	}
+	tokens := strings.Split(token, ",")
```

Comment:
What if usernames, passwords, or remotes include `,`, `@`, or `:`? Are we sufficiently sure that this is impossible in each case?

Reply:
We can't answer this yet. We need to make sure this is impossible and add it to the specs.

Reply:
Add a comment to the code to this effect for future readers